File: parsers/java/src/main/java/fluxdsl/json/FluxMapper.java
package fluxdsl.json;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import fluxdsl.FluxTree.Ast;
import fluxdsl.FluxTree.Item;
import fluxdsl.FluxTree.Value;
import fluxdsl.FluxTree.ObjEntry;
import java.util.ArrayList;

/** JSON serialization/deserialization bridge between {@code FluxTree} AST and {@code JsonAst} DTOs. */
public final class FluxMapper {
    private static final ObjectMapper MAPPER = new ObjectMapper()
            .enable(SerializationFeature.INDENT_OUTPUT)
            .findAndRegisterModules();

    private FluxMapper() {}

    /** Returns the preconfigured Jackson {@link ObjectMapper}. */
    public static ObjectMapper mapper() {
        return MAPPER;
    }

    /** Serializes an AST root to a JSON string. */
    public static String toJson(Ast.Root root) {
        try {
            return MAPPER.writeValueAsString(toDto(root));
        } catch (Exception e) {
            throw new RuntimeException("Serialization failed", e);
        }
    }

    /** Deserializes a JSON string back to an AST root. */
    public static Ast.Root fromJson(String json) {
        try {
            return fromDto(MAPPER.readValue(json, JsonAst.Root.class));
        } catch (Exception e) {
            throw new RuntimeException("Deserialization failed", e);
        }
    }

    /** Converts an AST root to a {@code JsonAst} DTO root. */
    public static JsonAst.Root toDto(Ast.Root root) {
        var items = new ArrayList<JsonAst.Item>();
        for (var item : root.items()) items.add(toDto(item));
        return new JsonAst.Root(items);
    }

    static JsonAst.Item toDto(Item item) {
        if (item instanceof Item.Include i) return new JsonAst.Item.Include(i.path());
        if (item instanceof Item.Schema s) return new JsonAst.Item.Schema(s.path());
        var p = (Item.Pair) item;
        return new JsonAst.Item.Pair(p.key(), toDto(p.value()));
    }

    static JsonAst.Value toDto(Value value) {
        if (value instanceof Value.Str s) return new JsonAst.Value.Str(s.value());
        if (value instanceof Value.Num n) return new JsonAst.Value.Num(n.value());
        if (value instanceof Value.Bool b) return new JsonAst.Value.Bool(b.value());
        if (value instanceof Value.Null) return new JsonAst.Value.Null();
        if (value instanceof Value.Obj o) {
            var entries = new ArrayList<JsonAst.ObjEntry>();
            for (var e : o.entries()) entries.add(toDto(e));
            return new JsonAst.Value.Obj(entries);
        }
        var l = (Value.Lst) value;
        var elements = new ArrayList<JsonAst.Value>();
        for (var e : l.elements()) elements.add(toDto(e));
        return new JsonAst.Value.Lst(elements);
    }

    static JsonAst.ObjEntry toDto(ObjEntry entry) {
        if (entry instanceof ObjEntry.ObjPair p)
            return new JsonAst.ObjEntry.Pair(p.key(), toDto(p.value()));
        if (entry instanceof ObjEntry.ObjInclude i) return new JsonAst.ObjEntry.Include(i.path());
        var s = (ObjEntry.ObjSchema) entry;
        return new JsonAst.ObjEntry.Schema(s.path());
    }

    /** Converts a {@code JsonAst} DTO root back to an AST root. */
    public static Ast.Root fromDto(JsonAst.Root dto) {
        var items = new ArrayList<Item>();
        for (var item : dto.items()) items.add(fromDto(item));
        return new Ast.Root(items);
    }

    static Item fromDto(JsonAst.Item dto) {
        if (dto instanceof JsonAst.Item.Include i) return new Item.Include(i.path());
        if (dto instanceof JsonAst.Item.Schema s) return new Item.Schema(s.path());
        var p = (JsonAst.Item.Pair) dto;
        return new Item.Pair(p.key(), fromDto(p.value()));
    }

    static Value fromDto(JsonAst.Value dto) {
        if (dto instanceof JsonAst.Value.Str s) return new Value.Str(s.value());
        if (dto instanceof JsonAst.Value.Num n) return new Value.Num(n.value());
        if (dto instanceof JsonAst.Value.Bool b) return new Value.Bool(b.value());
        if (dto instanceof JsonAst.Value.Null) return new Value.Null();
        if (dto instanceof JsonAst.Value.Obj o) {
            var entries = new ArrayList<ObjEntry>();
            for (var e : o.entries()) entries.add(fromDto(e));
            return new Value.Obj(entries);
        }
        var l = (JsonAst.Value.Lst) dto;
        var elements = new ArrayList<Value>();
        for (var e : l.elements()) elements.add(fromDto(e));
        return new Value.Lst(elements);
    }

    static ObjEntry fromDto(JsonAst.ObjEntry dto) {
        if (dto instanceof JsonAst.ObjEntry.Pair p)
            return new ObjEntry.ObjPair(p.key(), fromDto(p.value()));
        if (dto instanceof JsonAst.ObjEntry.Include i) return new ObjEntry.ObjInclude(i.path());
        var s = (JsonAst.ObjEntry.Schema) dto;
        return new ObjEntry.ObjSchema(s.path());
    }
}

File: parsers/python/fluxdsl/flux_mapper.py
from __future__ import annotations

import json

from .json_ast import (
    Root, Item, Pair, Include, Schema,
    Value, Str, Num, Bool, Null, Obj, Lst,
    ObjEntry, ObjPair, ObjInclude, ObjSchema,
    _to_dict, item_from_dict, value_from_dict, entry_from_dict, root_from_dict,
)

_PARSER_VALUE_TYPE: dict[str, str] = {
    "str": "string",
    "num": "number",
    "bool": "boolean",
    "null": "null",
    "obj": "object",
    "list": "list",
}

_DTO_VALUE_TYPE: dict[str, str] = {v: k for k, v in _PARSER_VALUE_TYPE.items()}


def _parser_value_to_dto(parser_val: dict) -> dict:
    t = parser_val["type"]
    dto_t = _DTO_VALUE_TYPE.get(t)
    if dto_t is None:
        raise ValueError(f"Unknown parser value type: {t}")
    d = {"type": dto_t}
    if t == "null":
        pass
    elif t == "object":
        entries = parser_val.get("entries", [])
        d["entries"] = [_parser_entry_to_dto(e) for e in entries]
    elif t == "list":
        elements = parser_val.get("elements", [])
        d["elements"] = [_parser_value_to_dto(e) for e in elements]
    else:
        d["value"] = parser_val["value"]
    return d


def _parser_entry_to_dto(parser_entry: dict) -> dict:
    t = parser_entry.get("type")
    if t == "include":
        return {"type": "include", "path": parser_entry["path"]}
    if t == "schema":
        return {"type": "schema", "path": parser_entry["path"]}
    return {"type": "pair", "key": parser_entry["key"], "value": _parser_value_to_dto(parser_entry["value"])}


def _dto_value_to_parser(dto_val: dict) -> dict:
    t = dto_val["type"]
    parser_t = _PARSER_VALUE_TYPE.get(t)
    if parser_t is None:
        raise ValueError(f"Unknown DTO value type: {t}")
    d = {"type": parser_t}
    if t == "null":
        pass
    elif t == "obj":
        entries = dto_val.get("entries", [])
        d["entries"] = [_dto_entry_to_parser(e) for e in entries]
    elif t == "list":
        elements = dto_val.get("elements", [])
        d["elements"] = [_dto_value_to_parser(e) for e in elements]
    else:
        d["value"] = dto_val["value"]
    return d


def _dto_entry_to_parser(dto_entry: dict) -> dict:
    t = dto_entry["type"]
    if t == "include":
        return {"type": "include", "path": dto_entry["path"]}
    if t == "schema":
        return {"type": "schema", "path": dto_entry["path"]}
    return {"key": dto_entry["key"], "value": _dto_value_to_parser(dto_entry["value"])}


def to_dto(parser_ast: list) -> Root:
    """Convert a parser AST list to a DTO Root with typed dataclasses."""
    items = []
    for node in parser_ast:
        t = node["type"]
        if t == "pair":
            items.append(Pair(key=node["key"], value=_dto_value_from_ast(node["value"])))
        elif t == "include":
            items.append(Include(path=node["path"]))
        elif t == "schema":
            items.append(Schema(path=node["path"]))
    return Root(items=items)


def _dto_value_from_ast(parser_val: dict) -> Value:
    t = parser_val["type"]
    if t == "string":
        return Str(value=parser_val["value"])
    if t == "number":
        return Num(value=parser_val["value"])
    if t == "boolean":
        return Bool(value=parser_val["value"])
    if t == "null":
        return Null()
    if t == "object":
        entries = [_dto_entry_from_ast(e) for e in parser_val.get("entries", [])]
        return Obj(entries=entries)
    if t == "list":
        elements = [_dto_value_from_ast(e) for e in parser_val.get("elements", [])]
        return Lst(elements=elements)
    raise ValueError(f"Unknown parser value type: {t}")


def _dto_entry_from_ast(parser_entry: dict) -> ObjEntry:
    if parser_entry.get("type") == "include":
        return ObjInclude(path=parser_entry["path"])
    if parser_entry.get("type") == "schema":
        return ObjSchema(path=parser_entry["path"])
    return ObjPair(key=parser_entry["key"], value=_dto_value_from_ast(parser_entry["value"]))


def from_dto(root: Root) -> list:
    """Convert a DTO Root back to a parser AST list."""
    result = []
    for item in root.items:
        if isinstance(item, Pair):
            result.append({"type": "pair", "key": item.key, "value": _ast_value_from_dto(item.value)})
        elif isinstance(item, Include):
            result.append({"type": "include", "path": item.path})
        elif isinstance(item, Schema):
            result.append({"type": "schema", "path": item.path})
    return result


def _ast_value_from_dto(val: Value) -> dict:
    if isinstance(val, Str):
        return {"type": "string", "value": val.value}
    if isinstance(val, Num):
        return {"type": "number", "value": val.value}
    if isinstance(val, Bool):
        return {"type": "boolean", "value": val.value}
    if isinstance(val, Null):
        return {"type": "null"}
    if isinstance(val, Obj):
        entries = [_ast_entry_from_dto(e) for e in val.entries]
        return {"type": "object", "entries": entries}
    if isinstance(val, Lst):
        elements = [_ast_value_from_dto(e) for e in val.elements]
        return {"type": "list", "elements": elements}
    raise ValueError(f"Unknown DTO value type: {type(val)}")


def _ast_entry_from_dto(entry: ObjEntry) -> dict:
    if isinstance(entry, ObjInclude):
        return {"type": "include", "path": entry.path}
    if isinstance(entry, ObjSchema):
        return {"type": "schema", "path": entry.path}
    return {"key": entry.key, "value": _ast_value_from_dto(entry.value)}


def to_json(root: Root) -> str:
    """Serialize a DTO Root to a JSON string."""
    return json.dumps(_to_dict(root), indent=2, ensure_ascii=False)


def from_json(json_str: str) -> Root:
    """Deserialize a JSON string to a DTO Root."""
    d = json.loads(json_str)
    return root_from_dict(d)
